Add DMB memory barrier to H2D transfer

daniestevez opened the pull request:
I still haven't tested this on hardware, so I'm marking it as a draft.

The generated code looks quite reasonable. Here is the relevant code compiled with `--release`:
```
000077b4 <xilinx_dma::axi_dma::AxiDma::start_h2d>:
    77b4:       e92d4070        push    {r4, r5, r6, lr}
    77b8:       e24dd010        sub     sp, sp, #16
    77bc:       e1a04002        mov     r4, r2
    77c0:       e1a06001        mov     r6, r1
    77c4:       e1a05000        mov     r5, r0
    77c8:       eb000934        bl      9ca0 <dmb>
    77cc:       e5950010        ldr     r0, [r5, #16]
    77d0:       e3a01a07        mov     r1, #28672      ; 0x7000
    77d4:       e59f2068        ldr     r2, [pc, #104]  ; 7844 <xilinx_dma::axi_dma::AxiDma::start_h2d+0x90>
    77d8:       e3a03004        mov     r3, #4
    77dc:       e5801004        str     r1, [r0, #4]
    77e0:       e08f2002        add     r2, pc, r2
    77e4:       e285100c        add     r1, r5, #12
    77e8:       e28d0004        add     r0, sp, #4
    77ec:       eb0050e9        bl      1bb98 <std::os::imp::unix::net::datagram::UnixDatagram::send>
    77f0:       e59d0004        ldr     r0, [sp, #4]
    77f4:       e3500001        cmp     r0, #1
    77f8:       15950010        ldrne   r0, [r5, #16]
    77fc:       13a01001        movne   r1, #1
    7800:       13811a07        orrne   r1, r1, #28672  ; 0x7000
    7804:       15801000        strne   r1, [r0]
    7808:       15950010        ldrne   r0, [r5, #16]
    780c:       15961010        ldrne   r1, [r6, #16]
    7810:       15801018        strne   r1, [r0, #24]
    7814:       15950010        ldrne   r0, [r5, #16]
    7818:       1580101c        strne   r1, [r0, #28]
    781c:       15950010        ldrne   r0, [r5, #16]
    7820:       15804028        strne   r4, [r0, #40]   ; 0x28
    7824:       13a00000        movne   r0, #0
    7828:       128dd010        addne   sp, sp, #16
    782c:       18bd8070        popne   {r4, r5, r6, pc}
    7830:       e59d0008        ldr     r0, [sp, #8]
    7834:       e59d100c        ldr     r1, [sp, #12]
    7838:       e28dd010        add     sp, sp, #16
    783c:       e8bd4070        pop     {r4, r5, r6, lr}
    7840:       eaffeced        b       2bfc <_ZN6anyhow5error31_$LT$impl$u20$anyhow..Error$GT$9construct17h3ef8ffe2f55ddd09E.llvm.18047369102374268922>
    7844:       0003712e        .word   0x0003712e

00009ca0 <dmb>:
    9ca0:       f57ff05f        dmb     sy
    9ca4:       e12fff1e        bx      lr
```

I think that the overhead of the function call is an acceptable prize to pay for the simplicity of this solution.

daniestevez committed:
add DMB memory barrier to H2D transfer

Following discussion on #2 a DMB memory barrier is added to H2D
transfers using a call to a C function.

bastibl commented:
Thanks for the PR! It looks like this is at the moment hard-coded for arm7. If possible, I'll give this a try on the MPSoC in the next week to test on arm8.

daniestevez commented:
Indeed. I hardcoded `-march=armv7-a` because it was using `-march=armv6` and complaining that there is no `dmb` instruction in that architecture. By the way, I'm building with `cross build --target=arm-unknown-linux-gnueabihf`. The `build.rs` might need some smartness to force the `-march` depending on the target.

daniestevez committed:
fix DMB C implementation to build multi-arch

The previous C implementation of DMB only built correctly for
arm. Now this builds for any arch by detecting whether arm or aarch64
is the target arch, and setting the C -march accordingly, or using
a nop function for any other arch (which can be useful when a
crate depends on xilinx-dma for some functionality, but has some
tests that could run under x86_64 because they don't use xilinx-dma).

daniestevez commented:
I've given this a clean-up and arranged the `build.rs` script so that it builds correctly for any arch. In `arm` I'm still forcing `-march=armv7-a`, but in `aarch64` I'm not specifying any CFLAGS and this builds correctly with `cross build --target=aarch64-unknown-linux-gnu`. I've included a nop implementation of the `dmb()` function to let this build under non-ARM archs.

I think this could be merged now, as I can't see any downsides.

bastibl commented:
Thanks for your work, very much appreciated. I am just testing it and wondered about the `march` flag. For me, the relevant ARM targets of `rustup target list` are:
```
>> rustup target list
aarch64-apple-darwin
aarch64-apple-ios
aarch64-apple-ios-sim
aarch64-fuchsia
aarch64-linux-android (installed)
aarch64-pc-windows-msvc
aarch64-unknown-linux-gnu (installed)
aarch64-unknown-linux-musl (installed)
aarch64-unknown-none
aarch64-unknown-none-softfloat
arm-linux-androideabi
arm-unknown-linux-gnueabi
arm-unknown-linux-gnueabihf
arm-unknown-linux-musleabi
arm-unknown-linux-musleabihf
armebv7r-none-eabi
armebv7r-none-eabihf
armv5te-unknown-linux-gnueabi
armv5te-unknown-linux-musleabi
armv7-linux-androideabi
armv7-unknown-linux-gnueabi
armv7-unknown-linux-gnueabihf
armv7-unknown-linux-musleabi
armv7-unknown-linux-musleabihf
armv7a-none-eabi
armv7r-none-eabi
armv7r-none-eabihf
```

I'm not sure, but it looks like this setting would force `arm` to `armv7`. I did a follow-up commit that changed the gating to check for `armv7` and `aarch64` and removed the `march` switch. I hope that's fine. I compile tested for `arm`, `armv7`, and `aarch64`.

Please let me know if this breaks things for you, then I'll revert.

bastibl commented:
I rebased and merged the commits without changes but GitHub doesn't seem to recognize :-/ I'll close the PR but commits are in `master`.

daniestevez commented:
Thanks! I hadn't noticed that there was both `arm-unknown-linux-gnueabihf` and `armv7-unknown-linux-gnueabihf`. It makes much more sense to build with `armv7` for the Zedboard, and it probably gives some improvements (I've noticed now a bunch of `dmb ish` in several Rust functions, where with `arm` there were none).

I tested the main branch including these changes and it seems that something doesn't work correctly. The call to `__dmb` doesn't end up in the binary (and neither the function itself), for instance when building with `cross build --target=armv7-unknown-linux-gnueabihf --release --examples`. Apparently the problem is that `armv7` still counts as `arch = arm` both from the perspective of the `CARGO_CFG_TARGET_ARCH` env variable and for the `#[cfg()]`'s. So I've had to changed these to `arm` to get the `__dmb` function when building for `armv7-unknown-linux-gnueabihf`. The problem with this change is that the build for `arm-unknown-linux-gnueabihf` fails. Probably there is a way to distinguish `arm` from `armv7`.

bastibl commented:
Arg -.- But now I'm pretty hopeful :-)

https://github.com/FutureSDR/xilinx-dma/commit/ebbe08e80ef46de426968ea61defb4e535f69916

Do you have a nice way to inspect cross-compiled assembler output? Looks like `cargo asm` doesn't work w/ cross-compilation through `cross`.

daniestevez commented:
Nice. I learned a few more things about cargo with your last commit.

I confirm that it works for `armv7-unknown-linux-gnueabihf`.

To look at the asm output I'm just using the `objdump` from my cross toolchain as `objdump -C -d`. At the moment I only have a cross toolchain for 32-bit ARM, so I can't really check the output for aarch64, but I trust that your last commit will work for it as well.

bastibl commented:
Great, thanks for testing. I looked at `aarch64` with `objdump` and it calls `__dmb` as expected. So I guess it's fine.
